### ISSUE TITLE
Wrong VarCouldBeVal report

### ISSUE THREAD

**Issue by EdwarDDay** [detekt/detekt#3805]
## Expected Behavior

I shouldn't get a report about `VarCouldBeVal` for a property in an object in a function.

## Observed Behavior

I get a `VarCouldBeVal` issue for a property, that cannot be changed to `val`

## Steps to Reproduce

The following code reproduces the issue
```kotlin
fun test(){
    val wrapper = object {
        var test: Boolean = true
    }
    wrapper.test = false
}
```

## Your Environment

* Version of detekt used: 1.17.0
* Version of Gradle used (if applicable): 7.0.2
* Gradle scan link (add `--scan` option when running the gradle task):
* Operating System and version:
* Link to your project (if it's a public repository): https://github.com/EdwarDDay/serialization.kprefs/pull/98

**Comment by cortinico** [comment]
I've done some initial investigation on this, and I was able to reproduce.

The problem is that the `VarCouldBeVal` rule is working only for local `var`. The rule is not able to track accesses to `var` outside of the local scope. Technically you could also `wrapper.test = false` in another file, and the rule will still not be able to detect that you re-assigned a `var`.

The current workaround is to suppress the rule locally.

We should discuss how to evolve this rule to use type-resolution and data-flow analysis to become more precise.

**Comment by AnoDest** [comment]
Also found same false-positive report for
```kotlin
    override var optionEnabled: Boolean
        get() = prefs.getBoolean(OPTION_ENABLED, false)
        set(value) = prefs.edit { putBoolean(OPTION_ENABLED, value) }
```

**Comment by waliahimanshu** [comment]
> as we can check if there is a setter declared at the declaration site.

Hi, can you explain what you mean by the above? Can we exclude `setters` from the config file ?

**Comment by cortinico** [comment]
> Can we exclude `setters` from the config file ?

This has been fixed in #3880. It's available since Detekt 1.18.0 + it now requires Type Resolution
